Add backend support for custom behaviors + validation endpoint

tw4l opened the pull request:
Backend support for #2151 

Adds support for specifying custom behaviors via a list of strings.

When workflows are added or modified, minimal backend validation is done to ensure that all custom behavior URLs are valid URLs (after removing the git prefix and custom query arguments).

A separate `POST /crawlconfigs/validate/custom-behavior` endpoint is also added, which can be used to validate a custom behavior URL. It performs the same syntax check as above and then:
- For URL directly to behavior file, ensures URL resolves and returns a 2xx/3xx status code
- For Git repositories, uses `git ls-remote` to ensure they exist (and that branch exists if specified)

tw4l committed:
Add backend support for custom behaviors

Includes very minimal validation at this point that any urls passed
(after stripping custom git syntax) successfully resolve.

tw4l committed:
Modify import order

tw4l committed:
Use functino scope for sample_crawl_data

tw4l committed:
Update sort filter tests for changes

tw4l committed:
Final fix for test counts

tw4l committed:
Add custom behaviors to org crawling defaults

tw4l committed:
Add custom behaviors to crawling defaults test

tw4l committed:
Switch test to use webrecorder.net

tw4l committed:
Limit validation fetch of custom behavior url to 10 seconds

tw4l committed:
Add custom behavior validation endpoint

- Change validation on crawlconfig add/update to only ensure URLs
have valid syntax
- Add dedicated endpoint that ensures URLs resolve and provides
more detail feedback on errors

tw4l committed:
Add validation success case to tests

tw4l committed:
Modify validation endpoint to accept one custom behavior url

tw4l committed:
Fix test URL typo

tw4l committed:
Add git repo validation via git ls-remote

tw4l committed:
Add pylint comment

tw4l committed:
Add validation of git branch

tw4l committed:
Remove unnecessary async def

tw4l commented:
@SuaYoo @ikreymer Git repo validation has been added using `git ls-remote` to check the repo exists and is reachable (not private or otherwise unable to connect), as well as that the branch exists if specified in the URL. Should be good to go now :)

SuaYoo reviewed (commented):
I'm getting a `500 Internal Server Error` when using the new validation endpoint for a plain URL that doesn't exist, like `https://nonexistenturl-for-testing-browsertrix.com`. I see a `200` as expected when using an existing URL like `https://webrecorder.net/`. A non-existent git repo correctly returns a `404`.

tw4l committed:
Gracefully handle aiohttp exceptions

tw4l commented:
> I'm getting a `500 Internal Server Error` when using the new validation endpoint for a plain URL that doesn't exist, like `https://nonexistenturl-for-testing-browsertrix.com`. I see a `200` as expected when using an existing URL like `https://webrecorder.net/`. A non-existent git repo correctly returns a `404`.

Thanks for testing! Fixed this and added a test case for URLs that won't resolve.

SuaYoo reviewed (approved):
Tested locally with frontend pointed to https://github.com/webrecorder/browsertrix/pull/2520, working well!

tw4l commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -1052,6 +1080,94 @@ async def re_add_all_scheduled_cron_jobs(self):
+            )
+
+        if branch:
+            time.sleep(2)
```
Maybe can reduce this? Was trying to avoid getting rate limited

```suggestion
            time.sleep(0.5)
```

ikreymer replied:
sure, that's probably ok

tw4l replied:
Replaced with `await asyncio.sleep(0.5)`

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -1052,6 +1080,94 @@ async def re_add_all_scheduled_cron_jobs(self):
+        """Validate git repository and branch, if specified, exist and are reachable"""
+        git_remote_cmd = ["git", "ls-remote", repo_url, "HEAD"]
+        try:
+            subprocess.run(git_remote_cmd, check=True)
```
I think we probably want to use the asyncio subprocess API here, eg:
https://docs.python.org/3/library/asyncio-subprocess.html

Should be fairly easy to switch, since we just want the exit code.

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -1052,6 +1080,94 @@ async def re_add_all_scheduled_cron_jobs(self):
+        is_git_repo = False
+        git_branch = ""
+
+        if url.startswith(GIT_PREFIX):
```
This seems like some duplication with `_validate_custom_behaviors_url_syntax()`, should it just call that, and maybe return something like `"git", url_parts` to avoid checks in multiple places? A bit of a nitpick

ikreymer commented on backend/btrixcloud/utils.py:
```diff
@@ -100,6 +101,15 @@ def is_falsy_bool(stri: Optional[str]) -> bool:
 
 
+def is_url(url: str) -> bool:
+    """Check if string is ad valid URL"""
```
```suggestion
    """Check if string is a valid URL"""
```

tw4l committed:
Switch subprocess to asyncio subprocess

tw4l committed:
Format with Black

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -1052,6 +1079,82 @@ async def re_add_all_scheduled_cron_jobs(self):
                     flush=True,
                 )
 
+    def _validate_behavior_git_repo(self, repo_url: str, branch: str = ""):
```
```suggestion
    async def _validate_behavior_git_repo(self, repo_url: str, branch: str = ""):
```

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -1052,6 +1079,82 @@ async def re_add_all_scheduled_cron_jobs(self):
+            raise HTTPException(status_code=400, detail="invalid_custom_behavior")
+
+        if is_git_repo:
+            self._validate_behavior_git_repo(url, branch=git_branch)
```
```suggestion
            await self._validate_behavior_git_repo(url, branch=git_branch)
```

tw4l committed:
Add missing async and await

tw4l committed:
Fix docstring typo

Co-authored-by: Ilya Kreymer <ikreymer@users.noreply.github.com>

tw4l committed:
Fix typo

tw4l committed:
Refactor to avoid duplication of url parsing

tw4l committed:
Use asyncio.sleep and reduce wait to 0.5 seconds

tw4l commented:
@ikreymer should be good to go now

ikreymer committed:
dockerfile: move git install as first step

ikreymer reviewed (approved):
Nice!
